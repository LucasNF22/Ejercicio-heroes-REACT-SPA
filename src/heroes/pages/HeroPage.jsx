import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers";

export const HeroPage = () => {

  const { id } = useParams();
  const navigate = useNavigate();  
  const hero = useMemo( () => getHeroById( id ), [ id ] );

  const onNavigateBack = () => {
    // const cardSlide = document.getElementById("1")
    // console.log(cardSlide);
    // cardSlide.classList.add("animate__animated animate__slideOutRight")
    navigate( -1 )
  }

  

  if( !hero ) {
    return <Navigate to="/marvel" /> 
  }

  return (
    // <div className="row mt-5">
    //   <div className="col-4">
    //     <img 
    //       src={`/assets/heroes/${ id }.jpg`} 
    //       alt={ hero.superhero } 
    //       className="img-thumbnail"
    //     />
    //   </div>
    //   <div  className="col-8">
    //     <h3>{ hero.superhero }</h3>

    //   </div>
    // </div>
    <article id="1" className="postcard img-thumbnail justify-self-center animate__animated animate__fadeIn ">
				
        <img className="postcard__img img-thumbnail animate__animated animate__flipInY" src={`/assets/heroes/${ id }.jpg`} alt="Image Title" />
				
        <div className="postcard__text">
				
          <h1 className="postcard__title">{ hero.superhero }</h1>
          
          <div className="small">				
            <div className="list-group-item mb-2"><b>Alter ego: </b>{ hero.alter_ego }</div>
            <div className="list-group-item mb-2"><b>Publisher: </b>{ hero.publisher }</div>
            <div className="list-group-item mb-2"><b>First appearance: </b>{ hero.first_appearance }</div>  	
          </div>

          <div className="postcard__bar"></div>
          <h5 className="mt-2">Characters:</h5>
          <p className="postcard__preview-txt">
            { hero.characters }
          </p>
          
          <button 
            className="btn btn-outline-primary"
            onClick={ onNavigateBack }
            
          >
            Volver
          </button>

			</div>
		</article>
  )
}
