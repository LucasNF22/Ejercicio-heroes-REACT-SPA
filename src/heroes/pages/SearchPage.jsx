import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import queryString from 'query-string';
import { getHeroesByName } from '../helpers';
import { HeroCard } from '../components/HeroCard';

export const SearchPage = () => {

  const navigate = useNavigate();
  const location = useLocation();
  
  const { q ='' } = queryString.parse( location.search );
  // console.log(queryString.parse(location.search));
  const heroes = getHeroesByName( q );

  const showSearch = ( q.length === 0 );
  const showError = ( q.length > 0 ) && heroes.length === 0;

  const { searchText, onInputChange } = useForm({
    searchText: q
  });



  const onSearchSubmit = (e) => {
    e.preventDefault();
    // if( searchText.trim().length <= 1 ) return;

    navigate(`?q=${ searchText }`)
  }


  return (
    <>
     <div className='animate__animated animate__fadeIn'>

      <h1>Encuentra tu Héroe</h1>
      <hr />
      
      <div className="container ">
        <div className="col-12">         
          <form onSubmit={ onSearchSubmit }>
            <input 
              type="text" 
              placeholder="Busca tu Héroe"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={ searchText }
              onChange={ onInputChange }
              />
            <button className="btn btn-outline-primary mt-2
            ">
              Encuentra tu Héroe
            </button>
          </form>
        </div>
        <div className="col-12">
          <h4 className='mt-4'>Resultados</h4>
          <hr />
          
          <div 
            className="alert alert-primary animate__animated animate__headShake"
            style={{ display: showSearch ? '' : 'none' }}
          >
            Busca un Héroe
          </div>
          
          <div 
            className="alert alert-danger animate__animated animate__headShake"
            style= {{ display: showError ? '' : 'none' }}
          >
            No hay resultados con <b>{ q }</b>
          </div>
          
          <div className='row rows-cols-1 row-cols-md-3 g-3 mt-1'>
            {
              heroes.map( hero => 
                <HeroCard key={hero.id} {...hero} />
                )
              }

          </div>
          
          
        </div>
      </div>
              </div>
    </>
  )
}
