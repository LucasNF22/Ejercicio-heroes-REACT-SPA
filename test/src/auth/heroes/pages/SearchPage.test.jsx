import { getByLabelText, getByText, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { SearchPage } from "../../../../../src/heroes/pages/SearchPage"


describe('Pruebas en <SearchPage />', () => {

    test('Debe mostrarse correctamente con valores por defecto', () => {
        
        const { container } = render(
            <MemoryRouter> 
                <SearchPage />  
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();        

    });
    
    test('Debe mostrar el resultado con el valor del input del queryString', () => {
        
        const { container } = render(
            <MemoryRouter initialEntries={['/search?q=batman']} > 
                <SearchPage />  
            </MemoryRouter>
        );

        const img = screen.getByRole('img');
        expect( img.src ).toContain( '/public/heroes/dc-batman.jpg' );
        
        const input = screen.getByRole('textbox');
        expect( input.value ).toBe( 'batman' );
        
        // screen.debug()

        const alertSearch = screen.getByLabelText('alert-search');
        const alertResults = screen.getByLabelText('alert-results');
            
        expect( alertSearch.style.display ).toBe('none');
        expect( alertResults.style.display ).toBe('none');
    });

    test('Debe mostrar error si no se encuentra el hero', () => {
        
        
    })
    
    test('Debe llamar al navigate a la nueva pantalla', () => {
        

    })

})