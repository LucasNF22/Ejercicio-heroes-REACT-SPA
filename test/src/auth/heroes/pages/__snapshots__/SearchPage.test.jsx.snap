// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> Debe mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    class="animate__animated animate__fadeIn"
  >
    <h1>
      Encuentra tu Héroe
    </h1>
    <hr />
    <div
      class="container "
    >
      <div
        class="col-12"
      >
        <form
          aria-label="form"
        >
          <input
            autocomplete="off"
            class="form-control"
            name="searchText"
            placeholder="Busca tu Héroe"
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary mt-2
 "
          >
            Encuentra tu Héroe
          </button>
        </form>
      </div>
      <div
        class="col-12"
      >
        <h4
          class="mt-4"
        >
          Resultados
        </h4>
        <hr />
        <div
          aria-label="alert-search"
          class="alert alert-primary animate__animated animate__headShake"
        >
          Busca un Héroe
        </div>
        <div
          aria-label="alert-results"
          class="alert alert-danger animate__animated animate__headShake"
          style="display: none;"
        >
          No hay resultados con 
          <b />
        </div>
        <div
          class="row rows-cols-1 row-cols-md-3 g-3 mt-1"
        />
      </div>
    </div>
  </div>
</div>
`;
